src/osp/hierarchical_bitset.h

```diff
@@ -0,0 +1,563 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <array>
+#include <bitset>
+#include <cstdint>
+#include <memory>
+#include <stdexcept>
+
+namespace osp
+{
+
+#ifdef __GNUC__
+
+    inline int ctz(uint64_t a) noexcept { return __builtin_ctzll(a); }
+
+#elif defined(_MSC_VER)
+
+    #include <intrin.h>
+
+    inline int ctz(uint64_t a) noexcept
+    {
+        unsigned long int b = 0;
+        _BitScanForward64(&b, a);
+        return b;
+    }
+#elif
+
+    static_assert(false, "Missing ctz 'count trailing zeros' implementation for this compiler");
+
+#endif
+
+/**
+ * @brief Divide and round up
+ */
+template<typename NUM_T, typename DENOM_T>
+constexpr decltype(auto) div_ceil(NUM_T num, DENOM_T denom) noexcept
+{
+    return (num / denom) + (num % denom != 0);
+}
+
+template<typename INT_T>
+constexpr bool bit_test(INT_T block, int bit)
+{
+    return (block & (INT_T(1) << bit) ) != 0;
```

Comment:
Looks to me like this function only works with integer types.

Recommend doing

static_assert(std::is_integral_v<INT_T>);

Reply:
Alternatively, you could just make this function take std::uintmax_t or std::intmax_t as it's argument.

Reply:
I'll go with static asserts. A possible problem with std::intmax_t, is that it can be bitshifted beyond the original type specified.

Reply:
Actually, It might be better to use std::is_unsigned_v, since bitshifting signed integers is undefined behaviour.

Reply:
:+1:

Reply:
Very petty, but I have a personal preference for writing literals in hex if they take part in bitwise ops, just as a signal that we're interested in the values' properties as machine primitives and not just abstract decimal integers. This is utterly trivial so you don't have to change it if you don't want, just thought I'd mention it in case you found it to be an interesting idiom. My way would simply look like:

```cpp
return (block & (INT_T(0x1) << bit) ) != 0x0;
```